rust/datafusion/src/execution/physical_plan/csv.rs

```diff
@@ -0,0 +1,194 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Execution plan for reading CSV files
+
+use std::fs;
+use std::fs::File;
+use std::sync::{Arc, Mutex};
+
+use crate::error::Result;
+use crate::execution::physical_plan::{BatchIterator, ExecutionPlan, Partition};
+use arrow::csv;
+use arrow::datatypes::{Field, Schema};
+use arrow::record_batch::RecordBatch;
+
+/// Execution plan for scanning a CSV file
+pub struct CsvExec {
+    /// Path to directory containing partitioned CSV files with the same schema
+    path: String,
+    /// Schema representing the CSV files after the optional projection is applied
+    schema: Arc<Schema>,
+    /// Does the CSV file have a header?
+    has_header: bool,
+    /// Optional projection for which columns to load
+    projection: Option<Vec<usize>>,
+    /// Batch size
+    batch_size: usize,
+}
+
+impl ExecutionPlan for CsvExec {
```

Comment:
Now thinking about it: would it be better if we rename `ExecutionPlan` something like `DataSource`?  Also with the current interface how can we pushdown various filters to the storage?

Reply:
A good reference may be Spark Data Source V2 APIs.

Reply:
This is more like `org.apache.spark.sql.execution.SparkPlan` which is the base class for physical operators.

Reply:
Thanks. The doc helped a lot, although this is different from other designs I've touched before (e.g., Hive, Impala) where there are query fragments and operators within a query fragment. The parallelism/repartitioning happens on query fragment level but instead here it is on the operator level.